WIP, ENH: SceneObject text overlay

tylerjereddy opened the pull request:
* Draft/demo for adding text overlay to a `SceneObject`. Currently uses 3D text objects per discussion at:
https://github.com/yuxuanzhuang/ggmolvis/issues/5#issuecomment-2568991896. Another possible approach is also mentioned there.

A sample render with the text overlay feature on this branch:

<details>
<summary> details: tiny sample render script for use with this branch </summary>

```python
import MDAnalysis as mda
from MDAnalysis.tests.datafiles import PSF, DCD
import ggmolvis
from ggmolvis.ggmolvis import GGMolVis


if __name__ == "__main__":
    ggmv = GGMolVis()
    u = mda.Universe(PSF, DCD)
    frame = u.trajectory.frame
    system = u.select_atoms("all")
    all_mol = ggmv.molecule(system)
    all_mol.set_text(f"{frame = }",
                     location=(-0.2, -0.7, 0.0))
    all_mol.render(resolution=(1000, 1000),
                   filepath="test_text_overlay.png")
```
</details>

![image](https://github.com/user-attachments/assets/872d0d87-f502-457f-abaa-77c72ae5e284)




PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

tylerjereddy committed:
WIP, ENH: SceneObject text overlay

* Draft/demo for adding text overlay to a `SceneObject`. Currently
uses 3D text objects per discussion at:
https://github.com/yuxuanzhuang/ggmolvis/issues/5#issuecomment-2568991896.
Another possible approach is also mentioned there.

tylerjereddy commented on ggmolvis/sceneobjects/base.py:
```diff
@@ -165,6 +165,20 @@ def set_color(self, color):
+        text_object = bpy.context.active_object
+        text_object.data.body = text
+        text_object.data.size = text_size
+        text_object.rotation_euler = rotation_radians
```
Some thoughts for design discussion after playing around with this for a bit:

1. This approach doesn't currently handle per-frame updating, though prototyping for the single-frame case made sense to me for a first pass.
2. This "3D text" approach seems to have the advantage of all kinds of fancy 3D object things like rotations and thickness of 3D objects and so on, but it also seems to have some typesetting drawbacks, like the alignment of the "=" symbol in the sample text above, which may not be a problem with other approaches. I imagine each approach will have its tradeoffs.
3. For this function/method, `location` was kind of set by trial and error--to be honest I often end up doing something similar when placing text in i.e., old-school VMD `tcl` scripts, but anchoring placement relative to "objects" of some sort might be interesting/useful.
4. Getting the correct rotation also took some trial and error, which was annoying.

BradyAJohnston commented:
Thanks for having a play around with this!

Some points re: your comments.

For updating, if this subclassed the `MolecularEntity` from MN, that provides a `self.set_frame()` method this is potentially called every time the frame is updated. This should then automatically handle updating per frame and also saving / restoring.

https://github.com/BradyAJohnston/MolecularNodes/blob/31aa1b96c17b67f30ce85dde9fcc2c0a6a3d87bd/molecularnodes/entities/base.py#L18

For the location, I don't know the exact math but I know it can be done (and likely helpers for it in the bpy API), but you can use the cameras transform to set the text's location in "screen space" so that location and rotation work relative to what the camera sees. This will be the only reliable approach for it

tylerjereddy commented:
@yuxuanzhuang also noted that I should probably be doing this work inside of the placeholder `class Text(SceneObject)`, which has built-in frame updating functionality.

There is also some question as to ecosystem duplication with `MolecularEntity`, but in any case prototyping with the placeholder is perhaps the first thing I'll try (and defer the discussion about which "backend" approach is best for now, and if they both need to exist).

tylerjereddy committed:
MAINT, ENH: PR 14 revisions

* Move text handling to the reviewer-suggested location, in the `Text`
class.

* Use `_update_frame` to update the frame number text in each frame.

tylerjereddy commented:
Ok, this is still a bit rough, but I've revised it based on Yuxuan's initial feedback, moving the infrastructure into the `Text` class. The frame numbers are now properly updating in the render--here is a sample video with the usual test ADK trajectory: https://youtube.com/shorts/H63-8RiZzSM?feature=share

The pertinent part of the base (i.e., user-facing) render script from https://github.com/tylerjereddy/render_north_star that required additions for this is shown in the diff below:

<details>

```diff
--- a/probe.py
+++ b/probe.py
@@ -2,7 +2,7 @@ import argparse
 
 
 import MDAnalysis as mda
-from ggmolvis.ggmolvis import GGMolVis
+from ggmolvis.ggmolvis import GGMolVis, Text
 import bpy
 import yaml
 
@@ -70,6 +70,10 @@ def main(p_config: dict):
                          background_color["green"],
                          background_color["blue"],
                          background_color["alpha"]))
+    text = Text(frame_counter="frame: ",
+                text_size=0.1,
+                rotation=(1.13, 0, 0),
+                location=(-0.2, -0.7, 0.0))
```

</details>

I don't like that the updating happens internally though, rather than via a user-provided variable, because that kind of corners us into offering a specific "updating option" rather than the flexibility of arbitrary updates based on user-provided math/variables. Maybe this is a step in that direction though.

The user needing to provide `location` and `rotation` for the text is not ideal either (unless they really want to), so investigating the camera-relative placements seems important still. I believe this may be messy while we have multiple cameras in the scene though, which is getting cleaned up elsewhere I think.

BradyAJohnston commented:
We could instead have the use provide a function which takes the frame as an integer (and maybe could expand to other values also) and produces the final text. That way the user has full control of creation of text and formatting.

PardhavMaradani commented:
Just a pointer. One way I was thinking about 3d text based annotations (alongside the possibilities with the `blf` module) was using Geometry Nodes itself - particularly starting with the `String to Curves` node. This has the advantage that there are no loose text objects in the scene, they are nicely tied to a single object/universe, have access to all the underlying geometry/attributes, they can be easily made to point to the camera all the time, etc. I didn't explore in detail, but just wanted to make a note since this is being explored.

BradyAJohnston commented:
There are two different ways to do the text approach, through Geometry Nodes and through a text object that has some kind of update function. 

While through Geometry Nodes is possible, I would say that the text manipulation through GN still isn't as mature enough as we would like for something like this, and exposing more control to the user through just using a text object rather than having object generation & transforms partially obfuscated inside of a node tree.

tylerjereddy committed:
BUG: PR 14 revisions

* `SceneObject` needs to set `self.lens` before
`_init_camera()`, otherwise initialization of the
`Text` class can fail.

tylerjereddy committed:
WIP, ENH: PR 14 revisions

* Initial attempt to place text overlay without
user provided `location` and `rotation`.

[ci skip]

tylerjereddy commented:
I've been experimenting a bit more here, to remove the need for the user to provide `location` and `rotation`. It isn't quite there yet, although some of the changes allow automatically rotating the text to face the camera. I think getting the translation done automatically is proving a bit annoying so far.

With just the text format and size specified by user:

```diff
+    text = Text(frame_counter="frame: ",
+                text_size=0.1)
```

The text gets placed in the scene, but not translated as desired yet (I'm intentionally extruding for now to make rotation issues more obvious when debugging):

![image](https://github.com/user-attachments/assets/c66e00d4-c23c-4566-8ad0-7e1f65e63091)

BradyAJohnston commented:
I've realised that we can probably do this a better way. Blender already has support for this kind of thing that we are doing. There area a series of "Object Constraints" you can apply to objects. If we do the `Copy Transforms`, selecting the camera object, then it will apply all of the transforms of the camera to the text object. If we change the object to apply these transformations _Before Original_ then we can transform the text object on the `Z` axis to push it out / in of the camera, then on the `XY` to move around in the camera space.

https://github.com/user-attachments/assets/7bf753e1-de9a-46dc-9cd6-116c804a233d

BradyAJohnston commented:
Code to set that up: 
```py
```py
obj = bpy.data.obejcts['Text']
con = obj.constraints.new("COPY_TRANSFORMS")
con.target = bpy.data.objects["Camera"]
con.mix_mode = 'BEFORE_FULL'
```

BradyAJohnston commented:
This would need to take into account changes in the zoom level of the camera though (focal length)

tylerjereddy committed:
WIP, ENH: PR 14 revisions

* Use constraints more correctly for frame count text placement.
Still needs work for getting the y coordinate of the text properly
adjusted as focal length changes.

[ci skip]

tylerjereddy commented:
It took some messing around to get the constraint to work properly with frame updating -- some parts of the logic seem to need to be in `_update_frame()`.

Once I figured that out, the main problem was indeed compensating automatically for focal length for: "scale" of text, y position of text in the view.

Getting the y position just right as focal length changes seems particularly annoying. Various LLM/internet searches favor transformation between 2D and 3D (world?) space involving camera matrices and stuff. Actually none of them gave me a reliable y coordinate that I could just set for my text object though, so I'm still left with a manual hack that doesn't work well.

I'd like to have a bulletproof/reliable way to get the `y` value I need for the text `location` as focal length varies. I'm not entirely convinced we can avoid the user having to make *some* decision here re: text scaling/size as focal length varies. This definitely seems messy.

BradyAJohnston commented:
If any Depth of Field in the camera is used as well, then this method will additionally not work as the text won't be on the focal plane.

yuxuanzhuang committed:
add frame counter

yuxuanzhuang committed:
add frame prop

yuxuanzhuang commented:
I’ve made some updates to demonstrate how I would use `Text` for annotations. The issue with using a **Text Object** is still present, but the basic idea is that Text should be extensible, allowing users to display arbitrary information and can be dynamic based on the input and the current frame.

The FrameCounter now can accept any frame number provider (with a `.frame` property) using
`frame_counter = FrameCounter(u.trajectory)` or `frame_counter = FrameCounter(gg_obj)`
